sled-agent/src/bootstrap/early_networking.rs

```diff
@@ -38,59 +44,185 @@ pub enum EarlyNetworkSetupError {
     Dendrite(String),
 
     #[error("Error during DNS lookup: {0}")]
-    DnsResolver(#[from] internal_dns::resolver::ResolveError),
+    DnsResolver(#[from] ResolveError),
 }
 
 /// Code for configuring the necessary network bits to bring up the control
 /// plane
-pub struct EarlyNetworkSetup {
-    log: Logger,
+pub struct EarlyNetworkSetup<'a> {
+    log: &'a Logger,
 }
 
-impl EarlyNetworkSetup {
-    pub fn new(log: &Logger) -> Self {
-        EarlyNetworkSetup { log: log.clone() }
+impl<'a> EarlyNetworkSetup<'a> {
+    pub fn new(log: &'a Logger) -> Self {
+        EarlyNetworkSetup { log }
+    }
+
+    pub async fn lookup_boundary_switch_addrs(
+        &self,
+        resolver: &DnsResolver,
+        config: &RackNetworkConfig,
+    ) -> HashSet<Ipv6Addr> {
+        let switch_zone_addrs = self.lookup_switch_zone_addrs(resolver).await;
+        let mut boundary_switch_addrs = HashSet::new();
+
+        for uplink_config in &config.uplinks {
+            if let Some(&zone_addr) =
+                switch_zone_addrs.get(&uplink_config.switch)
+            {
+                boundary_switch_addrs.insert(zone_addr);
+            } else {
+                // TODO-correctness is it okay to just return here if the config
```

Comment:
I think this is the right thing to do in this scenario. It's not a programmer error. It's configuration error, which could even be stale or the switch may not have come online yet, right?

Reply:
Hmmm! Now that you mention it, I think this function (more specifically, `map_switch_zone_addrs` which is called by `lookup_switch_zone_addrs`) will hang forever if one of the switches is offline, because it tries to resolve every switch zone IP returned by internal DNS, which should include both of them even if one of them is offline.

I'm not sure what the right thing to do here is to handle that.

Reply:
Attempting to lay out the problem explicitly, working backwards from where we end up using this information (the first few / lowest layers of which are unchanged on this PR, but I think already problematic):

1. In `ServiceManager::initialize_zone()`, we [call ServiceManager::opte_ports_needed()](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/services.rs#L1072).
2. In `ServiceManager::opte_ports_needed()`, we [build a list of DPD clients](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/services.rs#L875-L889), one for each IP addr in `request.boundary_switches`. (I'll come back to how `boundary_switches` gets populated below).
3. Still in `ServiceManager::opte_ports_needed()`, we send an [`ensure_nat_entry()`](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/services.rs#L936) request via each DPD client. This is wrapped in a [`backoff::retry_notify()`](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/services.rs#L959-L964), so `opte_ports_needed()` (and therefore `initialize_zone()`) will not return until it has successfully told DPD at all of the `request.boundary_switches` addresses about the NAT entry it needs.

IIUC, this means that starting any [external DNS, nexus, or NTP zone](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/services.rs#L864-L870) will block until it is able to talk to DPD on every boundary switch.

How do we populate the boundary switches? This is one of the major intended changes from #3729: on #3729, the boundary switch addresses are determined once at RSS time, and then saved in the bootstore. Absent any updates to the bootstore, that would mean any attempt to start an  external DNS/nexus/NTP zone would block until it successfully talks to DPD at the IP addresses of the switch zones that were present at RSS time. On this PR, we no longer store the boundary switch addresses in the bootstore; instead, we look them up whenever we need them:

1. During RSS, we look up the switch zone addresses both on:
  a. [first RSS run](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/rack_setup/service.rs#L922-L925), so we can [set `boundary_switches` on the initial zone requests](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/rack_setup/service.rs#L956-L958) and [send them to nexus](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/rack_setup/service.rs#L1024) (I have not looked into what Nexus does with these at all - it may warrant more investigation)
  b. [subsequent RSS runs that have already been applied](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/rack_setup/service.rs#L783-L785), in this case soley to [send them to nexus](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/rack_setup/service.rs#L794)
2. Any time sled-agent cold boots, it calls [load_services()](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/sled_agent.rs#L432-L440) to restart any services it's responsible for. This starts internal DNS first (if the sled-agent is responsible for internal DNS), then [looks up boundary switches](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/services.rs#L591-L607) so it can set `boundary_switches` for subsequent external DNS/nexus/NTP zones.
3. Any time sled-agent receives a request to start new services, it eventually lands in `ensure_all_services_persistent` which also [looks up boundary switches](https://github.com/oxidecomputer/omicron/blob/cold-boot-rss-uplink-config/sled-agent/src/services.rs#L2531-L2555) so it can set `boundary_switches` for any of the new zones it needs to start as part of the incoming request.

In all of these cases where we need to dynamically look up the boundary switches, we land in this early networking module and do the following:

1. Look up all dendrite services in internal DNS; this happens inside a backoff::retry and therefore cannot fail (but can block indefinitely if internal DNS is unavailable).
2. For every IPv6 addr returned by internal DNS in step 1, we ask MGS at that same address which switch it is. This happens inside an infinite retry loop and therefore cannot fail (but can block indefinitely if MGS is unavilable at any of these IP addresses).
3. In the non-RSS cases, we then filter the switch addresses down to just those that have uplinks present, based on the config provided at RSS time (which is stored in the bootstore). This cannot fail and does not block (we've already read the bootstore on sled-agent startup).

These latter three steps are technically infallible, but there are a bunch of ways they could block indefinitely, including at least:

* Internal DNS has stale addresses for the switch zones (we'd block forever trying to talk to a nonexistent MGS)
* One of the switches is down (we'd block until it comes back and we can talk to its MGS)

If we assume internal DNS is not stale _and_ it returns exactly two IP addresses, we technically only need a response from one: we could assume if the first one we hear back from is (wlog) switch 0, the other IP address must be switch 1, but that seems dicey:

* Assuming internal DNS is not stale without confirming we can actually talk to MGS at that address seems iffy
* If only one of the switch zones is up, even if we make assumptions that let us return, service startup is still going to block trying to talk to DPD at all the addresses we return

Reply:
> Internal DNS has stale addresses for the switch zones (we'd block forever trying to talk to a nonexistent MGS)

In a loop attempting to contact a switch zone, we should re-resolve each time to make things work in an eventually consistent setting.

> One of the switches is down (we'd block until it comes back and we can talk to its MGS)

This is where reconciliation semantics fit in quite well. We need to continuously try to drive toward the desired state of the world (2 switches with NAT entries in this case) without necessarily blocking a partially desired but still functional state (1 switch with NAT entries) from being realized.

Reply:
I'm resolving this conversation based on the in-person discussion; will post a separate followup with changes.